File: SourceCode/Python/ChalmersGU/ListMap.py

from API import Map
from KVPair import KVPair
from LinkedList import LinkedList

#/* *** ODSATag: ListMap *** */
#/* *** ODSATag: Header *** */
class LinkedMap(Map):
    def __init__(self):
        # This could also be a DynamicArrayList.
        self._internalList = LinkedList()
#/* *** ODSAendTag: Header *** */

#/* *** ODSATag: Put *** */
    def put(self, key, value):
        for entry in self._internalList:
            if key == entry.key:
                oldValue = entry.value
                entry.value = value
                return oldValue
        # If we're using a DynamicArrayList we should add at the end of the list instead.
        self._internalList.add(0, KVPair(key, value))
        return None
#/* *** ODSAendTag: Put *** */

#/* *** ODSATag: Get *** */
    def get(self, key):
        for entry in self._internalList:
            if key == entry.key:
                return entry.value
        return None
#/* *** ODSAendTag: Get *** */

#/* *** ODSATag: Remove *** */
    # Python iterators do not have a .remove() method,
    # so here is a slower version.
    def remove(self, key):
        for i, entry in enumerate(self._internalList):
            if key == entry.key:
                break
        else:
            return None
        removed = entry.value
        self._internalList.remove(i)
        return removed
#/* *** ODSAendTag: Remove *** */

#/* *** ODSATag: ContainsKey *** */
    def containsKey(self, key):
        return self.get(key) is not None
#/* *** ODSAendTag: ContainsKey *** */

#/* *** ODSATag: EmptySize *** */
    def isEmpty(self):
        return self._internalList.isEmpty()

    def size(self):
        return self._internalList.size()
#/* *** ODSAendTag: EmptySize *** */

#/* *** ODSATag: Iterator *** */
    def __iter__(self):
        for entry in self._internalList:
            yield entry.key
#/* *** ODSAendTag: Iterator *** */
#/* *** ODSAendTag: ListMap *** */


#######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed ##

def _printMap(m):
    print("[", " ".join(f"{k}:{m.get(k)}" for k in m), "]", m.size())

if __name__ == '__main__':
    m = LinkedMap()
    _printMap(m)
    print("Putting values")
    for i in range(40):
        m.put(chr((i%25)+65), i)
        if i % 5 == 0: _printMap(m)
    _printMap(m)
    print("Removing values")
    i = 7
    while not m.isEmpty():
        m.remove(chr((i%25)+65))
        i += 11
        if m.size() % 5 == 0: _printMap(m)

File: SourceCode/Java_Generic/ChalmersGU/ListMap.java

import java.util.Iterator;

/* *** ODSATag: ListMap *** */
/* *** ODSATag: Header *** */
class ListMap<K, V> implements Map<K, V> {
    private List<KVPair<K,V>> internalList;

    ListMap() {
        // This could also be a DynamicArrayList.
        internalList = new LinkedList<>();
    }
/* *** ODSAendTag: Header *** */

/* *** ODSATag: Put *** */
    public V put(K key, V value) {
        for (KVPair<K,V> entry : internalList) {
            if (key.equals(entry.key)) {
                V oldValue = entry.value;
                entry.value = value;
                return oldValue;
            }
        }
        // If we're using a DynamicArrayList we should add at the end of the list instead.
        internalList.add(0, new KVPair<>(key, value));
        return null;
    }
/* *** ODSAendTag: Put *** */

/* *** ODSATag: Get *** */
    public V get(K key) {
        for (KVPair<K,V> entry : internalList)
            if (key.equals(entry.key))
                return entry.value;
        return null;
    }
/* *** ODSAendTag: Get *** */

/* *** ODSATag: Remove *** */
    // To be able to remove entries efficiently,
    // the underlying list iterator has to implement the .remove() method.
    public V remove(K key) {
        Iterator<KVPair<K,V>> iter = internalList.iterator();
        while (iter.hasNext()) {
            KVPair<K,V> entry = iter.next();
            if (key.equals(entry.key)) {
                V removed = entry.value;
                iter.remove();   // Remove the last entry returned by the iterator.
                return removed;
            }
        }
        return null;
    }
/* *** ODSAendTag: Remove *** */

/* *** ODSATag: ContainsKey *** */
    public boolean containsKey(K key) {
        return get(key) != null;
    }
/* *** ODSAendTag: ContainsKey *** */

/* *** ODSATag: EmptySize *** */
    public boolean isEmpty() {
        return internalList.isEmpty();
    }

    public int size() {
        return internalList.size();
    }
/* *** ODSAendTag: EmptySize *** */

/* *** ODSATag: Iterator *** */
    public Iterator<K> iterator() {
        return new LinkedMapIterator();
    }

    private class LinkedMapIterator implements Iterator<K> {
        private Iterator<KVPair<K,V>> iter;
        LinkedMapIterator() {
            iter = internalList.iterator();
        }
        public boolean hasNext() {
            return iter.hasNext();
        }
        public K next() {
            return iter.next().key;
        }
    }
/* *** ODSAendTag: Iterator *** */
/* *** ODSAendTag: ListMap *** */


/***************************************************************************************/
/** What comes below is purely for debugging and testing purposes - it can be removed **/

    public void _printMap() {
        System.out.print("[ ");
        for (K k : this) System.out.print(k + ":" + get(k) + " ");
        System.out.println("] " + size());
    }

    public static void main(String[] args) {
        ListMap<String, Integer> map = new ListMap<>();
        map._printMap();
        System.out.println("Putting values");
        for (int i=0; i<40; i++) {
            map.put(String.valueOf((char)((i%25)+65)), i);
            if (i % 5 == 0) map._printMap();
        }
        map._printMap();
        System.out.println("Removing values");
        int i = 7;
        while (!map.isEmpty()) {
            map.remove(String.valueOf((char)((i%25)+65)));
            i += 11;
            if (map.size() % 5 == 0) map._printMap();
        }
    }
/* *** ODSATag: ListMap *** */
}
/* *** ODSAendTag: ListMap *** */
